### PR TITLE
Do not use `SourceText` indexer when parsing

### PR DESCRIPTION
**Pull request by Neme12**
Currently, `IsConflictMarkerTrivia` is the only place in the lexer that uses the indexer of `SourceText` as opposed to `PeekChar()` and other methods on `SlidingTextWindow`. Depending on the implementation of `SourceText`, indexing might not be optimal - that's why `SlidingTextWindow` even exists and why it copies the data from the `SourceText` into its own buffer. This matters especially for `StringBuilderText`, which is what's returned by `SyntaxNode.GetText()`, which is what source generators sometimes use to get a `SourceText` from a compilation root they built up using the syntax APIs.

This PR removes the usage of the indexer in the first commit. ~In the second commit, I enabled nullability in a few related files. In the 4th and 5th commits, I fixed some minor bugs related to conflict marker parsing that I noticed in the code. I recommend reviewing commit by commit.~
EDIT: I reverted the nullability changes and other changes to put them in separate PRs based on feedback.

Here's a benchmark showing the difference in calling `CSharpSyntaxTree.ParseText` on a `StringBuilderText` from Roslyn's Syntax.xml.Internal.Generated.cs file. The difference isn't huge, but it's noticable.
```
|   Method |                  Job |              Runtime |     Mean |   Error |  StdDev | Ratio | RatioSD |
|--------- |--------------------- |--------------------- |---------:|--------:|--------:|------:|--------:|
| ParseOld |             .NET 6.0 |             .NET 6.0 | 125.6 ms | 2.43 ms | 2.49 ms |  1.00 |    0.00 |
| ParseNew |             .NET 6.0 |             .NET 6.0 | 117.6 ms | 2.13 ms | 1.99 ms |  0.93 |    0.03 |
|          |                      |                      |          |         |         |       |         |
| ParseOld |        .NET Core 3.1 |        .NET Core 3.1 | 138.5 ms | 1.91 ms | 1.79 ms |  1.00 |    0.00 |
| ParseNew |        .NET Core 3.1 |        .NET Core 3.1 | 131.3 ms | 2.06 ms | 1.93 ms |  0.95 |    0.02 |
|          |                      |                      |          |         |         |       |         |
| ParseOld | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 143.8 ms | 2.87 ms | 2.82 ms |  1.00 |    0.00 |
| ParseNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 137.8 ms | 1.60 ms | 1.49 ms |  0.96 |    0.01 |
```

<details>
<summary>Source code of the benchmark</summary>

Note: for the purpose of this benchmark, I added a boolean flag to `Lexer` to switch between the old and the new implementation of `IsConflictMarkerTrivia` to be able to compare them in a single benchmark.

```c#
using System.Text;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.Text;

BenchmarkRunner.Run<ParseTextBenchmark>();

[SimpleJob(RuntimeMoniker.Net60)]
[SimpleJob(RuntimeMoniker.NetCoreApp31)]
[SimpleJob(RuntimeMoniker.Net472)]
public class ParseTextBenchmark
{
    private readonly SourceText _sourceText;

    public ParseTextBenchmark()
    {
        SourceText sourceText;

        using (var fileStream = File.OpenRead(@"C:\Users\{username}\source\roslyn\src\Compilers\CSharp\Portable\Generated\CSharpSyntaxGenerator\CSharpSyntaxGenerator.SourceGenerator\Syntax.xml.Internal.Generated.cs"))
            sourceText = SourceText.From(fileStream);

        var syntaxTree = CSharpSyntaxTree.ParseText(sourceText);
        _sourceText = syntaxTree.GetRoot().GetText(Encoding.UTF8);
    }

    [Benchmark(Baseline = true)]
    public SyntaxTree ParseOld()
    {
        return CSharpSyntaxTree.ParseText(
            _sourceText,
            options: null,
            path: "",
            diagnosticOptions: null,
            isGeneratedCode: null,
            toggleUseTextWindow: false,
            cancellationToken: default);
    }

    [Benchmark]
    public SyntaxTree ParseNew()
    {
        return CSharpSyntaxTree.ParseText(
            _sourceText,
            options: null,
            path: "",
            diagnosticOptions: null,
            isGeneratedCode: null,
            toggleUseTextWindow: true,
            cancellationToken: default);
    }
}
```

</details>

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs [hunk 1]
```diff
@@ -2422,34 +2421,31 @@ private void LexSyntaxTrivia(bool afterFirstToken, bool isTrailing, ref SyntaxLi
...
 
-            return false;
+            if (firstCh != '=' && TextWindow.PeekChar(s_conflictMarkerLength) != ' ')
+            {
+                return false;
+            }
+
+            // Keep the new line check last because TextWindow.Reset might need to recopy the buffer,
+            // although that's rare in practice.
+            return TextWindow.Position == 0 || SyntaxFacts.IsNewLine(getLastChar());
```
Comment (Neme12) [inline src/Compilers/CSharp/Portable/Parser/Lexer.cs:2442]: Really rare. At first, I kept this check at the top and tried debugging when parsing Roslyn's Syntax.xml.Internal.Generated.cs and a few other large files, and never hit the breakpoint in `TextWindow.Reset` where it had to copy stuff (and the benchmark results were the same). So I could copy the check back to the top if someone feels that is better.

**Comment by CyrusNajmabadi** [comment]
Can you extract the nrt work into its own PR?

**Comment by Neme12** [comment]
Ok, I'll have to rebase and do a force push then, give me a second.

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs [hunk 2]
```diff
@@ -2422,34 +2422,31 @@ private void LexSyntaxTrivia(bool afterFirstToken, bool isTrailing, ref SyntaxLi
 
         private bool IsConflictMarkerTrivia()
         {
-            var position = TextWindow.Position;
-            var text = TextWindow.Text;
-            if (position == 0 || SyntaxFacts.IsNewLine(text[position - 1]))
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/Lexer.cs:2427]: can you keep the logic the same, but just use the sliding window?  this seems more complex to understand.

Comment (Neme12) [reply src/Compilers/CSharp/Portable/Parser/Lexer.cs:2427]: The logic is the same. Which specific part do you feel is more complex?

File: src/Compilers/VisualBasic/Portable/CommandLine/CommandLineDiagnosticFormatter.vb [hunk 3]
```diff
@@ -78,35 +78,41 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
                 ' Builds leading spaces up to the error span
                 For position = Math.Min(sourceSpanStart, line.Start) To Math.Min(line.End, sourceSpanStart) - 1
+#Disable Warning RS0030 ' Do not used banned APIs - This isn't a parsing scenario.
```
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/CommandLine/CommandLineDiagnosticFormatter.vb:81]: this just seems to be excessive.  we're talking about tiny amounts of time.  We don't need to ban a totally reasonable api in a blanket fashion like this.  We can simply use traces to drive things.

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs [hunk 4]
```diff
@@ -2513,14 +2513,10 @@ private SyntaxListBuilder LexConflictMarkerDisabledText(ref SyntaxListBuilder tr
         private void LexConflictMarkerEndOfLine(ref SyntaxListBuilder triviaList)
         {
             this.Start();
-            while (SyntaxFacts.IsNewLine(this.TextWindow.PeekChar()))
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/Lexer.cs:2520]: similarly, please move to another PR.  i'm ok merging with the previous commit in that one.

**Comment by CyrusNajmabadi** [comment]
Done with pass.

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs [hunk 5]
```diff
@@ -2482,7 +2482,7 @@ private SyntaxListBuilder LexConflictMarkerDisabledText(ref SyntaxListBuilder tr
             while (true)
             {
                 var ch = this.TextWindow.PeekChar();
-                if (ch == SlidingTextWindow.InvalidCharacter)
+                if (ch == SlidingTextWindow.InvalidCharacter && TextWindow.IsReallyAtEnd())
```
[the remark this answers is not in the document]
Comment (Neme12) [reply src/Compilers/CSharp/Portable/Parser/Lexer.cs:2485]: How many PRs should I create? I wrote in the original post that different commits do different things and it's best to review commit by commit.

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/Lexer.cs:2427]: the logic before was to check that we were after a newline.  Now the logic checks the character we are on first.  I'd prefer to keep the old logic as it makes the most sense to me.  These markers must be after newlines, so that makes sense as the first thing to check.

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/Lexer.cs:2485]: > How many PRs should I create?

Either 2 or 3.  One PR for perf, and then either 1 or 2 prs for the fixes (i'm ok with teh fixes going in on one pr).

> I wrote in the original post that different commits do different things and it's best to review commit by commit.

Compiler PRs are squashed.  So it will all be one commit when it goes in.  As separate PRs though we can keep the changes cleanly separated.

File: src/Compilers/VisualBasic/Portable/CommandLine/CommandLineDiagnosticFormatter.vb (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Compilers/VisualBasic/Portable/CommandLine/CommandLineDiagnosticFormatter.vb:81]: If accessing 'Text' should be a problem there, then introduce a member that hides that, and have it either throw, or be obsolete so that statically you are stopped :)

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs [hunk 6]
```diff
@@ -1815,7 +1808,7 @@ private bool ScanIdentifier_SlowPath(ref TokenInfo info)
                                 // Debugger syntax allows @0x[hexdigit]+ for object address identifiers.
                                 if (info.IsVerbatim &&
                                     this.ModeIs(LexerMode.DebuggerSyntax) &&
-                                    (char.ToLower(TextWindow.PeekChar(1)) == 'x'))
+                                    (char.ToLowerInvariant(TextWindow.PeekChar(1)) == 'x'))
```
Comment (Neme12) [inline src/Compilers/CSharp/Portable/Parser/Lexer.cs:1811]: This surely must have been a mistake? I can't imagine a scenario where parsing would depend on the current culture.

**Comment by Neme12** [comment]
@RikkiGibson @jcouv Could I get another pair of eyes on this? Thanks.

File: src/Compilers/VisualBasic/Portable/CommandLine/CommandLineDiagnosticFormatter.vb (hunk 3 above)
Comment (Neme12) [reply src/Compilers/VisualBasic/Portable/CommandLine/CommandLineDiagnosticFormatter.vb:81]: Fixed in af8bc88d3c650dda244d8dc4e214e6a4e9120b27

**Comment by CyrusNajmabadi** [comment]
i can look at this next week :)

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Neme12) [reply src/Compilers/CSharp/Portable/Parser/Lexer.cs]: Removed.

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs (hunk 4 above)
Comment (Neme12) [reply src/Compilers/CSharp/Portable/Parser/Lexer.cs]: Removed.

**Comment by CyrusNajmabadi** [comment]
@Neme12 do you still want to talk this through?

**Comment by CyrusNajmabadi** [comment]
Moving ot draft for now. @Neme12 If you'd like to pick this back up again, let us know!
